### PR TITLE
Maker: Initialization waits for first sensor readings

### PR DESCRIPTION
**Pull request by islemaster**
Waits for the sound sensor, light sensor, and thermometer to each receive their initial data before the initialization promise resolves.  This adds about ~30ms to board initialization.

The original bug was described thus:

> Light / sound sensors don’t pick a value for the initial reading -> reads null then throws an error
> * Related: The first time you get temperature for tempSensor, you get a value like -439 degrees.  So wait until sensor is running before returning a value.
> * Related: Automatically start sensor when it’s first called (take out __Sensor.start() blocks)

You can repro these issues with this app: https://studio.code.org/projects/applab/wI7uZ5nAjEFm5WHG17boFA/view - change the boolean flags at the top of the file to explore different behaviors.  What they reveal is that requests for sensor values immediately after connecting fail, but data for all three sensors starts streaming right away and making the requests 20ms later succeeds about half the time.

Three solutions came to mind:

1. Make reading a sensor an interpreter-blocking operation.
2. Make board initialization wait for sensors to get initial values.
3. Explicitly request values for each sensor as part of the initialization process.

I went with option two.  Option one is attractive because it doesn't add additional delay to initialization, but we have no precedent for blocking property getters in the interpreter which is a puzzle I'd rather not solve right now, and even if I did solve it, lots of other sensor methods and getters refer to the value from native code so we could run into this problem from another direction.  I didn't think of three until after implementing two, but I'm not sure it's significantly different/better - I'm interested in your thoughts.

Unit testing was tricky - I don't feel like I've settled on a nice, clean way to simulate board responses yet.  I'm open to recommendations.

There's a lot of diff noise, mostly due to changing some methods that used to be synchronous to return promises.  That should show up cleanly in the commit list - I might recommend viewing the [first three commits](https://github.com/code-dot-org/code-dot-org/pull/14417/files/19c612eef00ac3eea0341ec877cc319ad540dbe4), and then [the last three commits](https://github.com/code-dot-org/code-dot-org/pull/14417/files/19c612eef00ac3eea0341ec877cc319ad540dbe4..ca34be0f919fa9bab661e446343f5c022b9d4f04).

### CODE REVIEW

File: apps/test/unit/lib/kits/maker/PlaygroundComponentsTest.js [hunk 1]
```diff
@@ -36,46 +36,52 @@ describe('Circuit Playground Components', () => {
...
+            // 'touchPad1',
+            // 'touchPad2',
+            // 'touchPad3',
+            // 'touchPad6',
+            // 'touchPad9',
+            // 'touchPad10',
+            // 'touchPad12',
+        ]);
+      });
     });
```
Comment (Hamms) [inline apps/test/unit/lib/kits/maker/PlaygroundComponentsTest.js:82]: worth adding an `onRejected` handler, here and below?

File: apps/test/unit/lib/kits/maker/PlaygroundComponentsTest.js [hunk 2]
```diff
@@ -36,46 +36,52 @@ describe('Circuit Playground Components', () => {
...
-      const componentsTwo = createCircuitPlaygroundComponents(boardTwo);
-      expect(componentsTwo.led.board === boardTwo).to.be.true;
+      return createCircuitPlaygroundComponents(boardOne)
+        .then(components => componentsOne = components)
+        .then(() => createCircuitPlaygroundComponents(boardTwo))
+        .then(components => componentsTwo = components)
+        .then(() => {
+          expect(componentsOne.led.board === boardOne).to.be.true;
+          expect(componentsTwo.led.board === boardTwo).to.be.true;
+        });
```
Comment (Hamms) [inline apps/test/unit/lib/kits/maker/PlaygroundComponentsTest.js]: Might be easier to structure this as a `Promise.all`

File: apps/test/unit/lib/kits/maker/PlaygroundComponentsTest.js [hunk 3]
```diff
@@ -84,7 +90,8 @@ describe('Circuit Playground Components', () => {
           let led;
 
           beforeEach(() => {
-            led = createCircuitPlaygroundComponents(board).colorLeds[pin];
+            return createCircuitPlaygroundComponents(board)
+              .then(({colorLeds}) => led = colorLeds[pin]);
```
Comment (Hamms) [inline apps/test/unit/lib/kits/maker/PlaygroundComponentsTest.js:112]: I thought if you had async components in a `beforeEach`, you needed to use the `done` method; is that not the case?

File: apps/test/unit/lib/kits/maker/PlaygroundComponentsTest.js [hunk 4]
```diff
@@ -36,46 +36,52 @@ describe('Circuit Playground Components', () => {
...
-          // 'touchPad0',
-          // 'touchPad1',
-          // 'touchPad2',
-          // 'touchPad3',
-          // 'touchPad6',
-          // 'touchPad9',
-          // 'touchPad10',
-          // 'touchPad12',
-      ]);
+      return createCircuitPlaygroundComponents(board).then(components => {
```
Comment (Hamms) [inline apps/test/unit/lib/kits/maker/PlaygroundComponentsTest.js:59]: are these returns doing anything?

File: apps/src/lib/kits/maker/PlaygroundComponents.js [hunk 5]
```diff
@@ -26,8 +26,12 @@ import Piezo from './Piezo';
  * @returns {Promise.<Object.<String, Object>>} board components
  */
 export function createCircuitPlaygroundComponents(board) {
-  return new Promise(resolve => {
-    resolve({
+  return Promise.all([
+    initializeSoundSensor(board),
+    initializeLightSensor(board),
+    initializeThermometer(board),
+  ]).then(sensors => {
```
Comment (Hamms) [inline apps/src/lib/kits/maker/PlaygroundComponents.js]: If you wanted, you could here do something like `.then(([soundSensor, lightSensor, thermometer]) => {`

File: apps/test/unit/lib/kits/maker/PlaygroundComponentsTest.js (hunk 3 above)
Comment (islemaster) [reply apps/test/unit/lib/kits/maker/PlaygroundComponentsTest.js:112]: You have two options: You can use the `done` method, or you can return a `Promise`.  We use the latter strategy here.

**Comment by islemaster** [comment]
I was having trouble getting the stub to fire the data event we need at the right moment, especially in tests where we're already using sinon to stub the clock.  I'll take another look at that though, see if I can clean it up.

File: apps/test/unit/lib/kits/maker/PlaygroundComponentsTest.js (hunk 1 above)
Comment (islemaster) [reply apps/test/unit/lib/kits/maker/PlaygroundComponentsTest.js:82]: Specifically, here's [the documentation for Mocha promise support](https://mochajs.org/#working-with-promises).  If the promise rejects the test will fail.

**Comment by islemaster** [comment]
Tests updated, PTAL.

Instead of timing-dependent events to resolve setup (even fake timing) I've stubbed `EventEmitter.prototype.once('data', callback)` to call its callback immediately, so now when we create these sensors in unit tests they're not async at all.  In the `PlaygroundComponents` tests, the sensors also get a sensible initial value.

This doesn't realistically test the 'wait for first data' behavior, but I suppose what I was doing before doesn't test that very well either, and at least this way the test is cleaner and more transparent about what's being stubbed.

**Pull request merged**